backend/cfg/cfg_regalloc_validate.ml

```diff
@@ -0,0 +1,845 @@
+    match t with
+    | Physical { location } ->
+      Format.fprintf ppf "R[%a]" Location.print location
+    | Named { stamp } ->
+      Format.fprintf ppf "%a" Printmach.reg
+        { Reg.dummy with raw_name; stamp; typ; spill }
+end
+
+module Register = struct
+  module For_print = struct
+    type t =
+      { raw_name : Reg.Raw_name.t;
+        stamp : int;
+        typ : Cmm.machtype_component;
+        spill : bool
+      }
+  end
+
+  type t =
+    { reg_id : Reg_id.t;
+      for_print : For_print.t
+    }
+
+  let create (reg : Reg.t) : t =
+    { reg_id = Reg_id.of_reg reg;
+      for_print =
+        { raw_name = reg.raw_name;
+          stamp = reg.stamp;
+          typ = reg.typ;
+          spill = reg.spill
+        }
+    }
+
+  let to_dummy_reg (t : t) : Reg.t =
+    { Reg.dummy with
+      raw_name = t.for_print.raw_name;
+      typ = t.for_print.typ;
+      stamp = t.for_print.stamp;
+      loc = Reg_id.to_loc_lossy t.reg_id
+    }
+
+  let print (ppf : Format.formatter) (t : t) : unit =
+    Reg_id.print ~typ:t.for_print.typ ~raw_name:t.for_print.raw_name
+      ~spill:t.for_print.spill ppf t.reg_id
+
+  let compare (t1 : t) (t2 : t) : int = Reg_id.compare t1.reg_id t2.reg_id
+
+  let equal (t1 : t) (t2 : t) : bool = compare t1 t2 = 0
+end
+
+module Instruction = struct
+  type 'a t =
+    { desc : 'a;
+      arg : Register.t array;
+      res : Register.t array
+    }
+
+  let to_prealloc (type a) ~(alloced : a instruction) (t : a t) : a instruction
+      =
+    { alloced with
+      arg = Array.map Register.to_dummy_reg t.arg;
+      res = Array.map Register.to_dummy_reg t.res
+    }
+end
+
+module Description = struct
+  type t =
+    { instructions : (int, basic Instruction.t) Hashtbl.t;
+      terminators : (int, terminator Instruction.t) Hashtbl.t
+    }
+
+  let make_instruction_helper t f instr =
+    f
+      ~is_regalloc_specific:
+        (match instr.desc with Op (Spill | Reload) -> true | _ -> false)
+      t.instructions instr
+
+  let make_terminator_helper t f instr =
+    f ~is_regalloc_specific:false t.terminators instr
+
+  let add_instr ~seen_ids ~is_regalloc_specific instructions instr =
+    let id = instr.id in
+    if Hashtbl.find_opt seen_ids id |> Option.is_some
+    then
+      Cfg_regalloc_utils.fatal
+        "Duplicate instruction no. %d while creating pre-allocation description"
+        id;
+    Hashtbl.add seen_ids id ();
+    if is_regalloc_specific
+    then
+      Cfg_regalloc_utils.fatal
+        "Instruction no. %d is specific to the regalloc phase while creating \
+         pre-allocation description"
+        id;
+    Hashtbl.add instructions id
+      { Instruction.desc = instr.desc;
+        arg = Array.map Register.create instr.arg;
+        res = Array.map Register.create instr.res
+      }
+
+  let create cfg =
+    Cfg_regalloc_utils.precondition cfg;
+    if Cfg_regalloc_utils.regalloc_debug
+    then
+      Cfg_with_layout.save_as_dot ~filename:"before.dot" cfg
```

Comment:
the function name has an index appended which is rather fragile, which is really annoying for writing testing scripts.
Do you think there is a way to compromise?

Reply:
The trailing stamp is indeed annoying, I was worried we
would get only one dot file in case there are several
failures. But we do not have a fallback mode so that is
essentially moot. Maybe just leave a CR-someday
mentioning a fallback mode.